### PR TITLE
Support batched service discoverer updates

### PR DESCRIPTION
**Pull request by NiteshKant**
__Motivation__

`ServiceDiscoverer#discover()` returns a `Publisher<ServiceDiscovererEvent>` where each `ServiceDiscovererEvent` contains the state of a single address. Clients always retry service discovery failures assuming transient failures but try to hold on to the last known addresses till a subsequent successful `discover()` call.  Upon a subsequent successful call; the client then has to decide after how many addresses should the old cache be disposed to reduce churn associated by deactivating an address and then subsequently activating again. Today this is done with heuristics around percentage address received between failed and successful discover call.

If we implement batch service discoverer events where a ServiceDiscoverer implementation can send all known addresses in one batch then the above mentioned heuristics can be removed.

__Modification__

- Modify `ServiceDiscoverer#discover()` to return a `Publisher<List<ServiceDiscovererEvent<ResolvedAddress>>>`
- Modify `DefaultServiceDiscoveryRetryStrategy` to enable/disable retention of addresses between retries as opposed to the percent based heuristics.

__Result__

Deterministic caching of addresses on service discoverer failures.

### CODE REVIEW

**Comment by NiteshKant** [comment]
Depends on #1144

File: servicetalk-client-api/src/main/java/io/servicetalk/client/api/ServiceDiscoverer.java [hunk 1]
```diff
@@ -47,5 +49,5 @@
      * </ul>
      * @return a {@link Publisher} that represents a stream of events from the service discovery system.
      */
-    Publisher<E> discover(UnresolvedAddress address);
+    Publisher<List<E>> discover(UnresolvedAddress address);
```
Comment (idelpivnitskiy) [inline servicetalk-client-api/src/main/java/io/servicetalk/client/api/ServiceDiscoverer.java:52]: Do we need a `List`? Will `Collection` be enough?

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultSingleAddressHttpClientBuilder.java [hunk 2]
```diff
@@ -263,8 +255,9 @@ public StreamingHttpClient buildStreaming() {
         // Track resources that potentially need to be closed when an exception is thrown during buildStreaming
         final CompositeCloseable closeOnException = newCompositeCloseable();
         try {
-            final SdStatusCompletable sdStatus = new SdStatusCompletable();
-            final Publisher<? extends ServiceDiscovererEvent<R>> sdEvents = ctx.discover(sdStatus);
+
+            final Publisher<ServiceDiscovererEvent<R>> sdEvents =
+                    ctx.serviceDiscoverer.discover(ctx.address()).flatMapConcatIterable(identity());
```
Comment (idelpivnitskiy) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultSingleAddressHttpClientBuilder.java:260]: If we change public API of SD to emit `List` of addresses, should we also change LB factory API to use the same API for `eventPublisher`? Looks a bit weird do have these API different.

**Review by idelpivnitskiy — COMMENTED** [review]
LGTM, 2 questions:

File: servicetalk-client-api/src/main/java/io/servicetalk/client/api/ServiceDiscoverer.java (hunk 1 above)
Comment (NiteshKant) [reply servicetalk-client-api/src/main/java/io/servicetalk/client/api/ServiceDiscoverer.java]: Good call, `Collection` is enough

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultSingleAddressHttpClientBuilder.java (hunk 2 above)
Comment (NiteshKant) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultSingleAddressHttpClientBuilder.java:260]: I think sending the `List` to load balancer does not have any value as LBs are not expected to cache, batch SD updates. Passing the `List` makes the API more complex, so I decided not to propagate the changes to the LB layer. If we see a need in the future, it should be straight forward to change.

**Comment by NiteshKant** [comment]
Failure attributed to: https://github.com/apple/servicetalk/issues/1148

@servicetalk-bot test this please

**Review by idelpivnitskiy — APPROVED** [review]
🚢

**Pull request merged**